src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs

```diff
@@ -224,19 +224,31 @@ public async IAsyncEnumerable<IReadOnlyList<LogLine>> WatchAsync([EnumeratorCanc
 
             using var _ = _logStreamCts.Token.Register(() => channel.Writer.TryComplete());
 
+            LogLine[]? backlogSnapshot = default;
+            var flushBacklogSync = new ManualResetEventSlim();
             void Log(LogLine log)
             {
-                channel.Writer.TryWrite(log);
+                if (flushBacklogSync.IsSet)
+                {
+                    channel.Writer.TryWrite(log);
+                    return;
+                }
+
+                flushBacklogSync.Wait(cancellationToken);
+                // We need to ensure we don't write this log to the channel if it was already in the backlog
+                if (backlogSnapshot?.Contains(log) == false)
```

Comment:
Can `backlogSnapshot` ever be `null` here? `flushBacklogSync.Set();` is called after `backlogSnapshot` is set. So if we are waiting for it to be set, `backlogSnapshot` should always be non-null.

Reply:
Correct. I had this throwing when null previously but that seemed extreme. Are you suggesting this is a case we should just bang the nullability warning away?

Reply:
I guess I personally would just do:

```csharp
Debug.Assert(backlogSnapshot is not null, "backlogSnapshot is always set before flushBacklogSync is set");
if (!backlogSnapshot.Contains(log))
```

(Note that this is really just a nit on the readability of the method. Not critical to change.)

Reply:
Yeah I personally like the `Debug.Assert` approach too but this seems like another example of where we're not super consistent in the repo.